azurelinuxagent/common/cgroups.py

```diff
@@ -0,0 +1,698 @@
+        cpu_delta = self.current_cpu_total - self.previous_cpu_total
+        system_delta = max(1, self.current_system_cpu - self.previous_system_cpu)
+
+        return round(float(cpu_delta * self.cgt.cpu_count * 100) / float(system_delta), 3)
+
+    def collect(self):
+        """
+        Collect and return a list of all cpu metrics. If no metrics are collected, return an empty list.
+
+        :return: [(str, str, float)]
+        """
+        self.update()
+        usage = self.get_cpu_percent()
+        return [("Process", "% Processor Time", usage)]
+
+
+class Memory(object):
+    def __init__(self, cgt):
+        """
+        Initialize data collection for the Memory hierarchy
+
+        :param cgt: CGroupsTelemetry
+        :return:
+        """
+        self.cgt = cgt
+
+    def update(self):
+        pass
+
+    def collect(self):
+        """
+        Collect and return a list of all memory metrics
+        """
+        self.update()
+        return []
+
+
+class CGroupsTelemetry(object):
+    """
+    Encapsulate the cgroup-based telemetry for the agent or one of its extensions, or for the aggregation across
+    the agent and all of its extensions. These objects should have lifetimes that span the time window over which
+    measurements are desired; in general, they're not terribly effective at providing instantaneous measurements.
+    """
+    _tracked = {}
+    _metrics = {
+        "cpu": Cpu,
+        "memory": Memory
+    }
+    _hierarchies = _metrics.keys()
+    was_tracked = ""
+    delete_empty_groups = False
+
+    @staticmethod
+    def hierarchies():
+        return CGroupsTelemetry._hierarchies
+
+    @staticmethod
+    def track_cgroup(cgroup):
+        """
+        Create a CGroupsTelemetry object to track a particular CGroups instance. Typical usage:
+        1) Create a CGroups object
+        2) Ask CGroupsTelemetry to track it
+        3) Tell the CGroups object to add one or more processes (or let systemd handle that, for its cgroups)
+
+        :param CGroups cgroup: The cgroup to track
+        """
+        name = cgroup.name
+        if not CGroupsTelemetry.is_tracked(name):
+            tracker = CGroupsTelemetry(name, cgroup=cgroup)
+            CGroupsTelemetry._tracked[name] = tracker
+
+    @staticmethod
+    def track_systemd_service(name):
+        """
+        Create the CGroups object for a systemd service and track it.
+
+        :param str name: Service name (without .service suffix) to be tracked.
+        """
+        service_name = "{0}.service".format(name)
+        if not CGroupsTelemetry.is_tracked(service_name):
+            cgroup = CGroups.for_systemd_service(service_name)
+            tracker = CGroupsTelemetry(service_name, cgroup=cgroup)
+            CGroupsTelemetry._tracked[service_name] = tracker
+
+    @staticmethod
+    def track_extension(name, cgroup=None):
+        """
+        Create all required CGroups to track all metrics for an extension and its associated services.
```

Comment:
...if the `name` is not already tracked.

Reply:
OK